keps/sig-storage/20200212-volume-group.md

```diff
@@ -0,0 +1,919 @@
+
+Use case 2:
+For some storage systems, volumes are always managed in a group. For these storage systems, they will have to create a group for a single volume if they need to implement a create volume function in Kubernetes. Providing a VolumeGroup API will be very convenient for them.
+
+Use case 3:
+Instead of taking individual snapshots one after another, VolumeGroup can be used as a source for taking a snapshot of all the volumes in the same volume group. This may be a storage level consistent group snapshot if the storage system supports it.  In any case, when used together with ExecutionHook, this group snapshot can be application consistent. For this use case, we will introduce another CRD GroupSnapshot.
+
+Use case 4:
+VolumeGroup can be used to manage group replication or consistency group replication if the storage system supports it. Note replication is out of scope for this proposal.  It is mentioned here as a potential future use case.
+
+Use case 5:
+VolumeGroup can be used to manage volume placement to either spread the volumes across storage pools or stack the volumes on the same storage pool. Related KEPs proposing the concept of storage pool for volume placement is as follows:
+  https://github.com/kubernetes/enhancements/pull/1353
+  https://github.com/kubernetes/enhancements/pull/1347
+We may not really need a VolumeGroup for this use case. A StoragePool is probably enough.  This is to be determined.
+
+Use case 6:
+VolumeGroup can also be used together with application snapshot. It can be a resource managed by the ApplicationSnapshot CRD.
+
+Use case 7: 
+Some applications may not want to use ApplicationSnapshot CRD because they don’t use Kubernetes workload APIs such as StatefulSet, Deployment, etc.  Instead, they have developed their own operators.  In this case it is more convenient to use VolumeGroup to manage persistent volumes used in those applications.
+
+### Goals
+
+* Provide an API to manage multiple volumes together in a group.
+* Provide an API to support consistency groups for snapshots, ensuring crash consistency across all volumes in the group.
+* Provide an API to take a snapshot of a group of volumes, not ensuring crash consistency.
+* Provide a design to facilitate volume placement using the group API (To be determined).
+* The group API should be generic and extensible so that it may be used to support other features in the future.
+* A VolumeGroup may potentially be used to support consistency group replication or group replication in the future, but providing design on replication group is not in the scope of this KEP. This can be discussed in the future.
+
+## Proposal for Consistency Groups and Group Snapshots
+
+This proposal introduces new CRDs VolumeGroup, VolumeGroupClass, and GroupSnapshot.
```

Comment:
I find it slightly counter-intuitive not using the `Volume` prefix on `GroupSnapshot`.

I'm assuming an end-user would care mostly about their `VolumeSnapshots` and `VolumeGroups`. Using either `VolumeGroupSnapshot` or `VolumeSnapshotGroup` would seem more logical with bias towards `VolumeGroupSnapshot`, because you care mostly about the `VolumeGroup` when creating `VolumeSnapshots` in reference to the `VolumeGroup`.

Reply:
I could add that prefix. I was just trying make the name shorter.